### PR TITLE
Add HOPS module

### PR DESCRIPTION
**Pull request by jfy133**
This adds a heatmap plot, as exported from the [HOPS](https://github.com/rhuebler/HOPS) postprocessing R script. This output does not have useable data for general stats columns, and therefore just makes the heatmap.

Please note that I have no experience in Python, so this is very likely ugly - but fortunately it is a simple module.

Given my lack-of-experience, it includes a few tweaks to the 'Coding with MultiQC' documentation which had me lost for a bit until I checked up other module. This was mainly on the structure of a module directory; I didn't know by reading the documents in what file which bit of example code was going in as some sentences included implicit knowledge. My rephrasing can definitely be improved, but this would require larger restructuring that may not sit with the docs concept so I preferred to not do that.

## If this PR is _not_ a new module
 - [ ] This comment contains a description of changes (with reason)
 - [ ] `CHANGELOG.md` has been updated
 - [ ] (optional but recommended): https://github.com/ewels/MultiQC_TestData contains test data for this change

## If this PR is for a new module
 - [x] There is example tool output for tools in the https://github.com/ewels/MultiQC_TestData repositor https://github.com/ewels/MultiQC_TestData/pull/155
 - [x] Code is tested and works locally (including with `--lint` flag)
 - [x] `CHANGELOG.md` is updated
 - [ ] `README.md` is updated
 - [x] `docs/README.md` is updated with link to below
 - [x] `docs/modulename.md` is created
 - [x] Everything that can be represented with a plot instead of a table is a plot
 - [x] Report sections have a description and help text (with `self.add_section`)
 - [x] There aren't any huge tables with > 6 columns (explain reasoning if so)
 - [x] Each table column has a different colour scale to its neighbour, which relates to the data (eg. if high numbers are bad, they're red)
 - [x] Module does not do any significant computational work

Example output: (zipped to allow upload to GH)
[multiqc_report.zip](https://github.com/ewels/MultiQC/files/4711042/multiqc_report.zip)

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,6 +6,9 @@
 
 #### New Modules
 
+* [**HOPS**]()
```
Comment (ewels) [inline CHANGELOG.md:9]:
```suggestion
* [**HOPS**](https://www.github.com/rhubler/HOPS)
```

File: docs/modules.md [hunk 2]
```diff
@@ -43,13 +43,15 @@ for any pull-requests.
 MultiQC modules are Python submodules - as such, they need their own
 directory in `/multiqc/` with an `__init__.py` file. The directory should
 share its name with the module. To follow common practice, the module
-code usually then goes in a separate python file (also with the same name)
-which is then imported by `__init__.py`:
+code itself usually then goes in a separate python file (also with the same
+name, i.e. `multiqc/<my_mod>/<my_mod>.py`) which is then imported by the
```
Comment (ewels) [inline docs/modules.md:47]:
```suggestion
name, i.e. `multiqc/modname/modname.py`) which is then imported by the
```

For consistency for the snippet below:

```python
from .modname import MultiqcModule
```

File: docs/modules.md [hunk 3]
```diff
@@ -112,12 +114,13 @@ Last but not least, remember to add your new module to the `CHANGELOG.md`,
...
-ended in `:MultiqcModule` - this tells MultiQC to try to execute a class or
-function called `MultiqcModule`.
+If you've copied one of the other entry point statements, it will have ended 
+in `:MultiqcModule` - this tells MultiQC to try to execute a class or function 
+called `MultiqcModule`.
 
 To use the helper functions bundled with MultiQC, you should extend this
-class from `multiqc.modules.base_module.BaseMultiqcModule`. This will give
+class from `multiqc.modules.base_module.BaseMultiqcModule` in your
+module code file (i.e. `multiqc/<my_mod>/<my_mod>.py`). This will give
```
Comment (ewels) [inline docs/modules.md:123]:
```suggestion
module code file (i.e. `multiqc/modname/modname.py`). This will give
```

File: docs/modules.md (hunk 2 above)
Comment (ewels) [reply docs/modules.md:47]: Thank you for improving the docs by the way! Not many do 🙇

File: docs/modules/hops.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
+---
+Name: HOPS
+URL: https://www.github.com/rhubler/HOPS
+Description: >
+    This tool performs screening of output from the ancient DNA optimised 
+    BLAST-replacement tool MALT, to identify taxa that have expected ancient 
+    DNA characteristics.
+
```
Comment (ewels) [inline docs/modules/hops.md:8]:
```suggestion
```

File: multiqc/modules/hops/hops.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        
+        # Convert JSON to dict for easier manipulation
+        for s in parsed_json: 
+            s_name = self.clean_s_name(s, f['root'])
+            if s_name in self.hops_data: 
+                log.debug("Duplicate sample name found! Overwriting: {}".format(s_name))
+            self.add_data_source(f, s_name=s_name)
+            self.hops_data[s_name] = {}
+            for t in parsed_json[s]:
+                self.hops_data[s_name][t] = parsed_json[s][t][0]
```
Comment (ewels) [inline multiqc/modules/hops/hops.py:60]: Regarding the `[0]` - I noticed that the test data had values in an array. Why is this? Can the files support more than one sample or something? Will that array ever have more than one value?

File: multiqc/modules/hops/hops.py [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+            'colstops': [
+                [1, '#ffffff'],
+                [2, '#ffff33'],
+                [3, '#ff7f00'],
+                [4, '#e41a1c'],
+            ],
+            'decimalPlaces': 0,
+            'legend': False,
+            'datalabels': False,
+            'ycats_samples': True,
```
Comment (ewels) [inline multiqc/modules/hops/hops.py:102]: `True` is the default..
```suggestion
            'xcats_samples': False,
```

File: multiqc/modules/hops/hops.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+            anchor = 'hops_heatmap',
+            description = '''
+            Heatmap of candidate taxa for downstream aDNA analysis, with 
+            intensity representing additive categories of possible 'positive' 
+            hits. Note that large numbers of samples can result in Y-axis label 
+            overlap, drag down to view all. 
+
+            Yellow: Edit Distance. 
+            Orange: Damage. 
+            Red: Edit Distance and Damage.
```
Comment (ewels) [inline multiqc/modules/hops/hops.py:116]:
```suggestion
            * Yellow: Edit Distance. 
            * Orange: Damage. 
            * Red: Edit Distance and Damage.
```

File: multiqc/modules/hops/hops.py (hunk 5 above)
Comment (jfy133) [reply multiqc/modules/hops/hops.py:60]: No it shouldn't. That was a leftover from the JSON parsing, where each value was in it's own list (for some reason). Will remove.
```suggestion
                self.hops_data[s_name][t] = parsed_json[s][t]
```

**Comment by jfy133** [comment]
> Quick read over the code looks good!

Wow! I'm surprised!

Small suggestion for the docs (given my comment in the Opening) - a small section at the beginning of the 'coding with multiqc' section listing of e.g. a small directory tree of the relevant files that you will be editing/creating would be useful (probably compressed to ignore stuff not edited,  or not relevent). But I dunno if that would fit. Basically to make it easier for someone to know what everything should look like (the docs are pretty good with flow, but particularly when I was putting stuff in `<module>.py` I got lost quite regularly.

Something like

```
multiqc
modules
│   ├── modules
│   │      └── <your_module> # create this
│   │         ├── <your_module>.py # create this
│   │         └──__init__.py # create this
│   └── utils
│       ├── config_defaults.yaml # edit this
│       ├── config.py
<...>

```

File: multiqc/modules/hops/hops.py (hunk 5 above)
Comment (ewels) [reply multiqc/modules/hops/hops.py]: Yeah, it's the reason that I was curious about really 😉 Just thinking ahead if there could be some other user turning up with a JSON file that has multiple values in that list.

But are you guys writing the tool as well? So I guess you have control over that?

Comment (jfy133) [reply multiqc/modules/hops/hops.py]: Yup, well at least I am peripherally involved - I added the JSON support. It's extremely unlikely that there could be any more then one value - the JSON is simply covered from a R matrix (rows are taxa, Columns are samples)

**Comment by ewels** [comment]
@jfy133 - I just pushed a minor tweak, making the colours a _little_ more muted 😎 and adding a little shine to the help text: only showing a warning if we have lots of samples and giving some nice background colours to the text describing the colours.

Please check you agree with my changes! 👍

**Comment by jfy133** [comment]
@ewels yup, good with me! Thanks!

**Pull request merged**